src/wms/src/main/java/org/geoserver/wms/map/RenderedImageMapOutputFormat.java

```diff
@@ -1030,13 +1031,11 @@ private RenderedImage directRasterRender(
                 // Get the reader
                 //
                 final Feature feature =
-                        mapContent
-                                .layers()
-                                .get(0)
-                                .getFeatureSource()
-                                .getFeatures()
-                                .features()
-                                .next();
+                        DataUtilities.first(
+                                mapContent.layers().get(0).getFeatureSource().getFeatures());
+                if (feature == null || feature.getProperty("grid") == null) {
+                    return null;
```

Comment:
The report indicated an unclear error was provided, did you wish to provide a clear error (or log warning) here?

Reply:
Questions:
- This also only applies to the fast past, did you need a similar check in streaming renderer?
- How does this work with geometry transformations, where raster is generated on the fly from feature content? Perhaps that is not on the fast path...

Reply:
I don't think there is a need to provide an error per se, because geoserver never does that when your styling doesn't produce any picture. There are many other cases when that can happen (for example a vector style for a raster layer, a style that has rules that somehow doesn't match with any of your data,...). If your WMS image is blank, I think it is the user's responsibility to double check if an appropriate style has been selected. But in this specific case you get an error instead of a stack trace and a strack trace that might only cause confusion.

And I guess this answers your other questions as well. In most cases, you will get a blank image, which is expected behaviour the way I see it. Only in this case you get the NPE.

Reply:
Okay blank image it is!